transports/onion/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+//! Tor based transport for libp2p. Connect through the Tor network to TCP listeners.
+//!
+//! Main entrypoint of the crate: [OnionTransport]
+//!
+//! ## Example
+//! ```no_run
+//! # use async_std_crate as async_std;
+//! # use libp2p_core::Transport;
+//! # async fn test_func() -> Result<(), Box<dyn std::error::Error>> {
+//! let address = "/dns/www.torproject.org/tcp/1000".parse()?;
+//! let mut transport = libp2p_onion::OnionAsyncStdNativeTlsTransport::bootstrapped().await?;
+//! // we have achieved tor connection
+//! let _conn = transport.dial(address)?.await?;
+//! # Ok(())
+//! # }
+//! # async_std::task::block_on(async { test_func().await.unwrap() });
+//! ```
+
+use std::sync::Arc;
+use std::{marker::PhantomData, pin::Pin};
+
+use address::{dangerous_extract_tor_address, safe_extract_tor_address};
+use arti_client::{DataStream, TorClient, TorClientBuilder};
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::{transport::TransportError, Multiaddr, Transport};
+use tor_rtcompat::Runtime;
+
+mod address;
+mod provider;
+
+#[cfg(feature = "tokio")]
+#[cfg_attr(docsrs, doc(cfg(feature = "tokio")))]
+#[doc(inline)]
+pub use provider::OnionTokioStream;
+
+use provider::OnionStream;
+
+pub type OnionError = arti_client::Error;
+
+#[derive(Clone)]
+pub struct OnionTransport<R: Runtime, S> {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<R>>,
+    /// The used conversion mode to resolve addresses. One probably shouldn't access this directly.
+    /// The usage of [OnionTransport::with_address_conversion] at construction is recommended.
+    pub conversion_mode: AddressConversion,
+    phantom: PhantomData<S>,
+}
+
+/// Configure the onion transport from here.
+pub type OnionBuilder<R> = TorClientBuilder<R>;
+
+/// Mode of address conversion. Refer tor [arti_client::TorAddr](https://docs.rs/arti-client/latest/arti_client/struct.TorAddr.html) for details.
+#[derive(Debug, Clone, Copy, Hash, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressConversion {
+    /// uses only dns for address resolution (default)
+    #[default]
+    DnsOnly,
+    /// uses ip and dns for addresses
+    IpAndDns,
+}
+
+impl<R: Runtime, S> OnionTransport<R, S> {
+    pub fn from_builder(
+        builder: OnionBuilder<R>,
+        conversion_mode: AddressConversion,
+    ) -> Result<Self, OnionError> {
+        let client = Arc::new(builder.create_unbootstrapped()?);
+        Ok(Self {
+            client,
+            conversion_mode,
+            phantom: PhantomData::default(),
+        })
+    }
+
+    pub async fn bootstrap(&self) -> Result<(), OnionError> {
+        self.client.bootstrap().await
+    }
+
+    pub fn with_address_conversion(&mut self, conversion_mode: AddressConversion) -> &mut Self {
```

Comment:
I am never quite sure what the Rust-idiomatic fn signature for builder functions is. I typically go for `self` because it allows usages like:

```rust
let client = OnionTransport::bootstrapped().await?.with_address_conversion(AddressConversion::IpAndDns);
```

Using `&mut self` forces an intermediary variable but I've had people [complain](https://github.com/testcontainers/testcontainers-rs/issues/382) about that in the past. Thoughts?

Reply:
Yes it's actually quite difficult and there is no really elegant solution in my opinion. Using Self is less flexible and requires assignments on the other hand it's more clean. I preferred `&mut self` because it allows for changing the conversion mode later on, while Self would be only really appropriate during creation. Since we already implement a builder pattern, or rather reexport one, `&mut self` would be more appropriate. Since `conversion_mode` is `pub` one could implement something custom if there is an issue.

Reply:
Yeah, there isn't really a usecase where you hold on to a `Transport` in libp2p. You construct them and pass them to the `Swarm`, at which point it is moved.